include/rmm/mr/cnmem_memory_resource.hpp

```diff
@@ -0,0 +1,141 @@
+ *---------------------------------------------------------------------------**/
+class cnmem_memory_resource final : public device_memory_resource {
+ public:
+  /**---------------------------------------------------------------------------*
+   * @brief Construct a cnmem memory resource and allocate the initial device
+   * memory pool
+
+   * TODO Add constructor arguments for other CNMEM options/flags
+   *
+   * @param initial_pool_size Size, in bytes, of the intial pool size. When
+   * zero, an implementation defined pool size is used.
+   *---------------------------------------------------------------------------**/
+  explicit cnmem_memory_resource(std::size_t initial_pool_size = 0) {
+    cnmemDevice_t dev;
+    // TODO Update exception
+    if (cudaSuccess != cudaGetDevice(&(dev.device))) {
+      throw std::runtime_error{"Failed to get CUDA device"};
+    }
+    // Note: cnmem defaults to half GPU memory
+    dev.size = initial_pool_size;
+    dev.numStreams = 1;
+    cudaStream_t streams[1];
+    streams[0] = 0;
+    dev.streams = streams;
+    dev.streamSizes = 0;
+    unsigned flags = 0;
+    // TODO Update exception
+    auto status = cnmemInit(1, &dev, flags);
+    if (CNMEM_STATUS_SUCCESS != status) {
+      std::string msg = cnmemGetErrorString(status);
+      throw std::runtime_error{"Failed to intialize cnmem: " + msg};
+    }
+  }
+
+  ~cnmem_memory_resource() {
+    auto status = cnmemFinalize();
+#ifndef NDEBUG
+    if (status != CNMEM_STATUS_SUCCESS) {
+      std::cerr << "cnmemFinalize failed.\n";
+    }
+#endif
+  }
+
+  bool supports_streams() const noexcept override { return true; }
+
+ private:
+  /**---------------------------------------------------------------------------*
+   * @brief Allocates memory of size at least \p bytes using cnmem.
+   *
+   * The returned pointer has at least 256B alignment.
+   *
+   * @throws `std::runtime_error` if cnmem failed to register the stream
+   * @throws `std::bad_alloc` if the requested allocation could not be fulfilled
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @return void* Pointer to the newly allocated memory
+   *---------------------------------------------------------------------------**/
+  void* do_allocate(std::size_t bytes, cudaStream_t stream) override {
+    register_stream(stream);
+    void* p{nullptr};
+    auto status = cnmemMalloc(&p, bytes, stream);
+    if (CNMEM_STATUS_SUCCESS != status) {
+#ifndef NDEBUG
+      std::cerr << "cnmemMalloc failed\n";
+#endif
+      throw std::bad_alloc{};
+    }
+    return p;
+  }
+
+  /**---------------------------------------------------------------------------*
+   * @brief Deallocate memory pointed to by \p p.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   *---------------------------------------------------------------------------**/
+  void do_deallocate(void* p, std::size_t, cudaStream_t stream) override {
```

Comment:
Why does do_deallocate() take a size? It's common at the point of deallocation to not know the size of the allocation -- allocators maintain this information for you.

Reply:
This is what [`std::pmr::memory_resource`](https://en.cppreference.com/w/cpp/memory/memory_resource/deallocate) does. I did consider deviating, but here were my thoughts on why I chose to keep it:

- I believe the reason `std::pmr::memory_resource::deallocate` requires a `size` argument is to allow for simpler, and therefore more efficient, allocators. If the resource doesn't need to track how large your allocation was, then that simplifies the implementation of the allocator and therefore can make it more efficient. 
- Furthermore, it should be rare to call `deallocate` directly. Instead, it should be done in some RAII wrapper (like `device_buffer`) that already tracks the size of its allocation and therefore it's no extra burden to require calls to `deallocate` to pass the size.

Reply:
Understood. I think we just need clarify what if any of RMM this is planning to replace. If RMM owns the default memory resource, and rmmAlloc is still a thing, then rmmAlloc has to start tracking allocations in order to provide a size to the memory resource for deallocation.

If rmmAlloc is going away in favor of device_buffer, then there is no issue.